ddtrace/internal/wrapping/context.py

```diff
@@ -18,6 +19,117 @@
 
 T = t.TypeVar("T")
 
+# ============================================================================
+# Lazy wrapping machinery: allows deferring expensive bytecode instrumentation
+# until the first time a function is actually called.
+# ============================================================================
+
+_lazy_registry: t.Dict[CodeType, "LazyMeta"] = {}
+
+
+class LazyMeta:
+    """Metadata for a lazily-wrapped function."""
+
+    __slots__ = ("func", "builder", "lock", "initialized")
+
+    def __init__(self, func: FunctionType, builder: t.Callable[[FunctionType], CodeType]):
+        self.func = func  # the original function object
+        self.builder = builder  # callable: builder(func) -> CodeType
+        self.initialized = False
+
+
+def __lazy_trampoline_entry(*args, **kwargs):
+    """
+    Called from the trampoline code on first invocation.
+    Uses the current frame's code object to find the right LazyMeta,
+    builds the heavy bytecode, and swaps it in place.
+    """
+    # Get the caller's frame (the trampoline frame)
+    code = sys._getframe(1).f_code
+    meta = _lazy_registry[code]
+
+    if not meta.initialized:
+        if not meta.initialized:
+            # Build heavy code from the original pre-wrapped function
+            new_code = meta.builder(meta.func)
+            # Swap the code on the SAME function object frameworks already hold
+            meta.func.__code__ = new_code
+            meta.initialized = True
+        # Drop the registry entry to free memory
+        _lazy_registry.pop(code, None)
+
+    # Now call the function again; this time it runs the heavy code
+    return meta.func(*args, **kwargs)
+
+
+def _make_trampoline_code(template_code: CodeType) -> CodeType:
```

Comment:
We can probably re-use existing technology that is hopefully generic enough to cover all cases, including freevars
```python
from ddtrace.internal.wrapping import wrap


def test_wrap_self_mutation():
    cell = " is the answer"

    def target(a, b):
        nonlocal cell

        return a + b + cell

    def wrapper(f, args, kwargs):
        def mutated(a, b):
            nonlocal cell

            return b + a + cell

        f.__code__ = mutated.__code__
        return f(*args, **kwargs)

    assert target("2", "4") == "24 is the answer"

    wrap(target, wrapper)

    assert target("2", "4") == "42 is the answer"
```